### PR TITLE
Disallow magic variables in inline assembly

### LINKED ISSUE #1144: Inline assembly: cannot use this
```
contract A {
    function a() returns (uint a) {
        assembly {
            a := extcodesize(this)
        }
    }
}
```

`this` is a valid reference when used with `extcodesize` / `balance` / etc, but it doesn't deposit a value to the stack:

```
Untitled1:5:18: Error: Expected instruction(s) to deposit 1 item(s) to the stack, but did deposit 0 item(s).
            a := extcodesize(this)
                 ^---------------^
```

### PR DESCRIPTION
**Pull request by axic**
Fixes #1144.

### CODE REVIEW

File: test/libsolidity/InlineAssembly.cpp [hunk 1]
```diff
@@ -157,6 +157,13 @@ BOOST_AUTO_TEST_CASE(oversize_string_literals)
 	BOOST_CHECK(!successAssemble("{ let x := \"123456789012345678901234567890123\" }"));
 }
 
+BOOST_AUTO_TEST_CASE(magic_variables)
+{
+	BOOST_CHECK(!successAssemble("{ this }"));
+	BOOST_CHECK(!successAssemble("{ ecrecover }"));
+	BOOST_CHECK(successAssemble("{ let ecrecover := 1 ecrecover }"));
```
Comment (yann300) [inline test/libsolidity/InlineAssembly.cpp:169]: I'm not sure but this test should fail no?

File: Changelog.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ Features:
 Bugfixes:
  * Disallow unknown options in `solc`
  * Inline assembly: support the `address` opcode
+ * Inline assembly: disallow magic variables (`this`, `super`, etc.)
```
Comment (axic) [inline Changelog.md]: I think this should be reworded. What really happens is that _magic variables_ are not given as identifiers to inline assembly and there can be used as local identifiers.

File: test/libsolidity/InlineAssembly.cpp (hunk 1 above)
Comment (axic) [reply test/libsolidity/InlineAssembly.cpp:169]: Actually no, it is properly. Added a comment to clarify the changelog.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -592,6 +592,10 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 			unsigned pushes = 0;
 			if (dynamic_cast<FunctionDefinition const*>(declaration))
 				pushes = 1;
+			else if (dynamic_cast<MagicVariableDeclaration const*>(declaration))
+			{
+				return false;
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Wouldn't the proper fix be to put `else return false;` at the end (and / or call `fatalTypeError()`)?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: Changelog.md (hunk 2 above)
Comment (axic) [reply Changelog.md]: Does this sound better?

> Inline assembly: magic variables (`this`, `super`, etc.) are not visible in inline assembly and they can be used as local variables names

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: Yeah probably that is simpler :)

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -608,7 +608,8 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 				if (!contract->isLibrary())
 					return false;
 				pushes = 1;
-			}
+			} else
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: `else` should go on its own line.

File: Changelog.md [hunk 5]
```diff
@@ -9,6 +9,8 @@ Features:
 Bugfixes:
  * Disallow unknown options in `solc`
  * Inline assembly: support the `address` opcode
+ * Inline assembly: magic variables (`this`, `super`, etc.) are not visible
+   in inline assembly and they can be used as local variables names
```
Comment (chriseth) [inline Changelog.md]: Everything that is not a keyword can be used as local variable names, they just shadow.

I think the actual fix here is that external variables of unsupported types are properly detected.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: Changelog.md (hunk 5 above)
Comment (axic) [reply Changelog.md]: Would you mind writing the text here for the changelog?

Comment (axic) [reply Changelog.md]: Added some rewording.

**Comment by axic** [comment]
@chriseth is this accepted now?

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
@axic @yann please review my changes to the changelog.

File: Changelog.md [hunk 6]
```diff
@@ -2,17 +2,16 @@
...
- * Disallow unknown options in `solc`
- * Inline assembly: support the `address` opcode
+ * Disallow unknown options in ``solc``
+ * Inline assembly: support the ``address`` opcode
  * Inline assembly: fix parsing of assignment after a label.
- * Inline assembly: external variables of unsupported type (such as `this`, `super`, etc.)
-   are properly detected. They are not available in inline assembly and can be used as
-   local variable names.
+ * Inline assembly: external variables of unsupported type (such as ``this``, ``super``, etc.)
+   are properly detected as unusable.
```
Comment (axic) [inline Changelog.md:16]: I think it is useful to mention that they are indeed usable as locals.

**Pull request merged**